### PR TITLE
Avoid retrying non-idempotent RPCs in binary connections (#549)

### PR DESCRIPTION
**Pull request by csringhofer**
See the #549 for the detailed analyses of the issue.

The fix works similarly to the existing solution for http connections:
- each RPC knows whether it is idempotent
- if the error comes from establishing the connection, then retry
- if the error comes from executing the RPC, only retry if the RPC is idempotent

A test is added that relies on slow metadata handling in the Impala cluster to trigger timouts. It would be nice to add wider and more reliable tests in the future similarly to the http tests in test_hs2_fault_injection.py

### CODE REVIEW

File: impala/tests/test_dbapi_connect.py [hunk 1]
```diff
@@ -252,6 +252,44 @@ def test_https_connection(self):
...
 
+    def test_retry_dml(self):
+        """Regression test for #549."""
+        # Connection with higher timeout to avoid errors.
+        self.connection = connect(ENV.host, ENV.port, timeout=10)
+        # Connect with low timeout to trigger failed RPCs.
+        # TODO: This reproduces the issue in the default Impala dev env (no local catalog),
+        #       but may fail to reproduce it in the future if Impala will get faster.
+        #       Ideally the error would be injected in a more stable way.
+        low_timeout_connection = connect(ENV.host, ENV.port, timeout=3)
```
Comment (rizaon) [inline impala/tests/test_dbapi_connect.py:263]: Create a constant for both short and long timeout value.

**Comment by joemcdonnell** [comment]
I think this fix makes sense.

It looks like impala-shell uses entirely different retry logic, and it limits retries to HTTP clients: https://github.com/apache/impala/blob/master/shell/impala_client.py#L692-L696

File: impala/hiveserver2.py [hunk 2]
```diff
@@ -1144,41 +1144,45 @@ def __init__(self, client, retries=3):
...
-                log.exception('Failed to open transport (tries_left=%s)',
-                              tries_left)
-                last_http_exception = None
-            except TTransportException:
-                log.exception('Failed to open transport (tries_left=%s)',
-                              tries_left)
-                last_http_exception = None
+            except socket.error as e:
+                if open_finished and not safe_to_retry: raise e
+                msg = "RPC failed" if open_finished else "Failed to open transport"
```
[the remark this answers is not in the document]
Comment (csringhofer) [reply impala/hiveserver2.py]: Instead of creating a nested function merged the expect blocks,

**Comment by csringhofer** [comment]
> I think this fix makes sense.
> 
> It looks like impala-shell uses entirely different retry logic, and it limits retries to HTTP clients: https://github.com/apache/impala/blob/master/shell/impala_client.py#L692-L696

Looked a bit into whether the non-http retries can be actually useful.
TLDR: for non-SSL sockets it seems useful, while for SSL sockets it only helps with idempotent RPCs

The code tries to ensure that the transport is working with
def open_transport(transport):
    if not transport.isOpen():
        transport.open()

When using raw socket, isOpen() peeks the socket (since Thrift 0.14.0) https://github.com/apache/thrift/blob/cb9ceada554f47aa5ebbedfe3984de0983cf0226/lib/py/src/transport/TSocket.py#L92 
So it can detect if the connection is broken and reestablish it with transport.open(). One caveat is that open() will throw TTransportException.ALREADY_OPEN in this case, as the failed isOpen() doesn't close the transport. (I will fix this in the next patch set)
Not sure if it is really useful to have retry loop in this case - if transport.open() threw an error, then retrying it without any sleep will probably also fail.

In case of SSLSocket currently isOpen won't check the underlying socket, see 
https://issues.apache.org/jira/browse/THRIFT-5595 . This was improved in Thrift 0.17.0 (Impyla still uses 0.16.0), but my impression is that it can still return false positive, so in that case that error will be detected during the actual Thrift RPC.

If there is no reliable way to peek SSLSockets, then I think that a sane workaround would be to do a cheap RPC without side-effect before doing a non-idempotent RPC. The time since the last RPC could be checked, e.g. only ping impalad if there was no RPC in the last few seconds.

File: impala/tests/test_dbapi_connect.py [hunk 3]
```diff
@@ -236,22 +238,62 @@ def test_jwt_auth_with_ldap_password(self):
...
+        cur.execute(create)
+        successful_inserts = 0
+        NUM_INSERTS = 5
+        for i in range(NUM_INSERTS):
+            try:
+                cur.execute(invalidate)
+                cur.execute(insert)
+                successful_inserts += 1
+            except:
+                con = connect(ENV.host, ENV.port, timeout=3)
```
Comment (rizaon) [inline impala/tests/test_dbapi_connect.py:283]: Replace `timeout=3` with `LOW_TIMEOUT_S`

File: impala/tests/test_dbapi_connect.py [hunk 4]
```diff
@@ -236,22 +238,62 @@ def test_jwt_auth_with_ldap_password(self):
...
 
+    def test_retry_dml(self):
+        """Regression test for #549."""
+        # Connection with higher timeout to avoid errors.
+
+        self.connection = connect(ENV.host, ENV.port, timeout=TIMEOUT_S)
+        # Connect with low timeout to trigger failed RPCs.
+        # TODO: This reproduces the issue in the default Impala dev env (no local catalog),
+        #       but may fail to reproduce it in the future if Impala will get faster.
+        #       Ideally the error would be injected in a more stable way.
```
Comment (rizaon) [inline impala/tests/test_dbapi_connect.py:265]: Just curious, can you utilize `debug_action` option to inject fault?

File: impala/hiveserver2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rizaon) [reply impala/hiveserver2.py]: ack.

File: impala/tests/test_dbapi_connect.py (hunk 4 above)
Comment (rizaon) [reply impala/tests/test_dbapi_connect.py:265]: Otherwise, code looks OK.

**Review by rizaon — APPROVED** [review]

**Comment by csringhofer** [comment]
Thanks for the reviews!
